compiler/rustc_lint/src/late.rs

```diff
@@ -337,10 +329,8 @@ impl<'tcx, T: LateLintPass<'tcx>> hir_visit::Visitor<'tcx> for LateContextAndPas
         hir_visit::walk_path(self, p);
     }
 
-    fn visit_attribute(&mut self, hir_id: hir::HirId, attr: &'tcx ast::Attribute) {
-        self.with_lint_attrs(hir_id, |cx| {
-            lint_callback!(cx, check_attribute, attr);
-        })
+    fn visit_attribute(&mut self, attr: &'tcx ast::Attribute) {
+        lint_callback!(self, check_attribute, attr);
```

Comment:
Can you give a more specific example? I know this change fixes the performance problem and passes all the tests, but I don't know enough about linting to know if there is some edge case that it will miss.

Reply:
Only clippy uses late lints on attributes.  They should probably made early lints, and drop support for late lints on attributes.

For instance:
```rust
#[warn(macro_use_imports)]
#[macro_use]
extern crate std;
```
and with swapped attributes.

Reply:
> Is attribute order supposed to be significant? Are attributes supposed to apply to other attributes on the same item, or do they just apply to the item? The latter makes more sense to me, but I honestly don't know.
> 

For now, the order of attributes is not significant.  My question is mostly: are we changing behaviour?

> > They should probably made early lints, and drop support for late lints on attributes.
> 
> Sorry, I don't understand this. Is this an argument in favour of this PR's change, or against it, or is it orthogonal?
> 

This is rather orthogonal. I'll post a follow-up issue if you don't mind.

> I tried your example, with this code (including slight necessary changes):
> 
> ```rust
> #[warn(clippy::macro_use_imports)]
> #[macro_use]
> extern crate std as other_std;
> 
> fn main() { }
> ```
> 
> With a standard rustc (not containing this PR's changes) I get much the same output from `cargo clippy` regardless of the order of the attributes:
> 
> ```
> warning: unused `#[macro_use]` import
>  --> src/main.rs:2:1
>   |
> 2 | #[macro_use]
>   | ^^^^^^^^^^^^
>   |
>   = note: `#[warn(unused_imports)]` on by default
> 
> error: useless lint attribute
>  --> src/main.rs:1:1
>   |
> 1 | #[warn(clippy::macro_use_imports)]
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: if you just forgot a `!`, use: `#![warn(clippy::macro_use_imports)]`
>   |
>   = note: `#[deny(clippy::useless_attribute)]` on by default
>   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_attribute
> ```
> 
> If I add the `!` as suggested then the error goes away, unsurprisingly.

So, IIUC, the behaviour does not change.  Great! Let's go ahead with this PR.

Reply:
Excellent! Do you want me to do anything else, or is it ready for an r+?